### ISSUE TITLE
Weird bug (missing triangles) in 3 dimensions

### ISSUE THREAD

**Issue by torvalamo** [mapbox/earcut#92]
All vertices in these examples are on a unit sphere, and the bug only occurs (seemingly randomly) when the polygon has vertices with both positive and negative z. 

    <- [0.2721957230595233,  -0.9605350502397119, -0.05728791852648584,
        0.27076767353958436, -0.9623895143093251,  0.022165056133977172,
        0.3083526498204863,  -0.9486345806634813,  0.0707889519494777,
        0.3619110094397141,  -0.9318858643761064, -0.02468110658629935]
    -> [3, 0, 1]

    <- [0.14763130978955374, -0.9886623805588933, -0.027417031886090803,
        0.16465209361109379, -0.9848327078391015,  0.054719517905281886,
        0.1681491534395761,  -0.9836098460481434,  0.06509633595450239,
        0.2241703576575745,  -0.9744326147750235,  0.015123822611557114,
        0.19173701906136723, -0.9794193836609019, -0.06304432116186201] 
    -> [1, 0, 4, 3, 2, 1]

    <- [0.05599310106571683, -0.997853082904166,  -0.03397053976162725,
        0.06641092297498565, -0.9977911752752119, -0.0015361811487453257,
        0.16465209361109379, -0.9848327078391015,  0.054719517905281886,
        0.14763130978955374, -0.9886623805588933, -0.027417031886090803]
    -> [0, 3, 2]

The added image shows closeup of some of the holes. It's a voronoi diagram of points on the surface of the sphere, and the tetrahedrons mark the points. For those who don't know what a voronoi diagram is, I'll just explain what is supposed to happen: The top missing triangle should be brown (belong to the blue point), the middle should be blue (belong to the cyan point) and the bottom should be purple (belong to the purple point).

You can see into the middle of the sphere where there is an axis indicator (+x=red, +y=green, +z=blue). These holes only occur when z=~0, but strangely not for all such cases (eg. the white point with the green/gray polygon)

The three examples above correspond to these exact holes and the bug is reproducable ad nauseum with those points.

![screen shot 2017-12-31 at 03 14 27](https://user-images.githubusercontent.com/397369/34458595-c9a3e084-edd8-11e7-9c3b-69f413959c2a.png)

**Comment by torvalamo** [comment]
Found another one, where the missing triangle doesn't cross the z=0, but the polygon itself does.

    <- [0.25386886620570015, 0.9645782029271353,  0.07169023091949686,
        0.1683845752448163,  0.9848687153560296, -0.04099083229926776,
        0.16941381468947908, 0.984395253684814,  -0.047591426908317344,
        0.19929329953952615, 0.9782798946332323, -0.057014283430037435,
        0.22307460312050684, 0.9741902161692462, -0.03451295644748336,
        0.28359320291907103, 0.958230386414503,   0.03700569969632519,
        0.26677060611712194, 0.96135773793472,    0.06800546614016607,
        0.25528670680233934, 0.9641233412761631,  0.07276592702978064]
    -> [0, 7, 6, 6, 5, 4, 0, 6, 4, 0, 4, 3, 1, 0, 3]

Missing triangle should be (i think) [1, 2, 3] in some order.

![screen shot 2017-12-31 at 03 58 30](https://user-images.githubusercontent.com/397369/34458795-595bd8d4-eddf-11e7-841d-455f3f19a412.png)

**Comment by torvalamo** [comment]
I experimented some more, and it seems this is what https://github.com/mapbox/earcut/issues/86 is about. If I reverse the input array, the triangulation is correct. So I'm guessing it's just a coincidence with the voronoi library outputting the polygons in reverse order in some cases where it crosses the z axis.

For future reference, I bypassed the problem with

    t = earcut(v, null, 3);
    if ( t.length / 3 !== ((v.length / 3) - 2) t = earcut(v.reverse(), null, 3);

simply redoing the calculation with a reverse order if the result size doesn't match.

**PS: This only works if the polygon is convex, as the number of triangles is then predetermined.**
